File: objc/main/test/objc/TestInterptimeErrors.m
#import "TestInterptimeErrors.h"

@implementation UserHiddenName

- (id) initWithName:(NSString *)name {
  if ( (self=[super init]) != nil ) {
    name = name;
  }
  return self;
}

- (NSString *) getName {
  return name;
}

- (void) dealloc {
  [name release];
  [super dealloc];
}

@end

@implementation UserHiddenNameField

- (id) initWithName:(NSString *)name {
  if ( (self=[super init]) != nil ) {
    name = name;
  }
  return self;
}

- (void) dealloc {
  [name release];
  [super dealloc];
}

@end

@implementation TestInterptimeErrors

- (void) testMissingEmbeddedTemplate {
  ErrorBuffer *errors = [ErrorBuffer newErrorBuffer];
  NSString *templates = [@"t() ::= \"<foo()>\"" stringByAppendingString:Misc.newline];
  [self writeFile:tmpdir arg1:@"t.stg" arg2:templates];
  STGroup *group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"t.stg"]] autorelease];
  [group setListener:errors];
  ST *st = [group getInstanceOf:@"t"];
  [st render];
  NSString *expected = [@"context [t] 1:0 no such template: foo" stringByAppendingString:newline];
  NSString *result = [errors description];
  [self assertEquals:expected arg1:result];
}

- (void) testMissingSuperTemplate {
  ErrorBuffer *errors = [ErrorBuffer newErrorBuffer];
  NSString *templates = [@"t() ::= \"<super.t()>\"" stringByAppendingString:Misc.newline];
  [self writeFile:tmpdir arg1:@"t.stg" arg2:templates];
  STGroup *group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"t.stg"]] autorelease];
  [group setListener:errors];
  NSString *templates2 = [@"u() ::= \"blech\"" stringByAppendingString:Misc.newline];
  [self writeFile:tmpdir arg1:@"t2.stg" arg2:templates2];
  STGroup *group2 = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"t2.stg"]] autorelease];
  [group importTemplates:group2];
  ST *st = [group getInstanceOf:@"t"];
  [st render];
  NSString *expected = [@"context [t] 1:1 no such template: super.t" stringByAppendingString:newline];
  NSString *result = [errors description];
  [self assertEquals:expected arg1:result];
}

- (void) testNoPropertyNotError {
  ErrorBuffer *errors = [ErrorBuffer newErrorBuffer];
  NSString *templates = [@"t(u) ::= \"<u.x>\"" stringByAppendingString:Misc.newline];
  [self writeFile:tmpdir arg1:@"t.stg" arg2:templates];
  STGroup *group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"t.stg"]] autorelease];
  [group setListener:errors];
  ST *st = [group getInstanceOf:@"t"];
  [st add:@"u" arg1:[[[User alloc] init:32 arg1:@"parrt"] autorelease]];
  [st render];
  NSString *expected = @"";
  NSString *result = [errors description];
  [self assertEquals:expected arg1:result];
}

- (void) testHiddenPropertyNotError {
  ErrorBuffer *errors = [ErrorBuffer newErrorBuffer];
  NSString *templates = [@"t(u) ::= \"<u.name>\"" stringByAppendingString:Misc.newline];
  [self writeFile:tmpdir arg1:@"t.stg" arg2:templates];
  STGroup *group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"t.stg"]] autorelease];
  [group setListener:errors];
  ST *st = [group getInstanceOf:@"t"];
  [st add:@"u" arg1:[[[UserHiddenName alloc] init:@"parrt"] autorelease]];
  [st render];
  NSString *expected = @"";
  NSString *result = [errors description];
  [self assertEquals:expected arg1:result];
}

- (void) testHiddenFieldNotError {
  ErrorBuffer *errors = [ErrorBuffer newErrorBuffer];
  NSString *templates = [@"t(u) ::= \"<u.name>\"" stringByAppendingString:Misc.newline];
  [self writeFile:tmpdir arg1:@"t.stg" arg2:templates];
  STGroup *group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"t.stg"]] autorelease];
  [group setListener:errors];
  ST *st = [group getInstanceOf:@"t"];
  [st add:@"u" arg1:[[[UserHiddenNameField alloc] init:@"parrt"] autorelease]];
  [st render];
  NSString *expected = @"";
  NSString *result = [errors description];
  [self assertEquals:expected arg1:result];
}

- (void) testSoleArg {
  ErrorBuffer *errors = [ErrorBuffer newErrorBuffer];
  NSString *templates = [@"t() ::= \"<u({9})>\"\n" stringByAppendingString:@"u(x,y) ::= \"<x>\"\n"];
  [self writeFile:tmpdir arg1:@"t.stg" arg2:templates];
  STGroup *group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"t.stg"]] autorelease];
  [group setListener:errors];
  ST *st = [group getInstanceOf:@"t"];
  [st render];
  NSString *expected = [@"context [t] 1:1 passed 1 arg(s) to template u with 2 declared arg(s)" stringByAppendingString:newline];
  NSString *result = [errors description];
  [self assertEquals:expected arg1:result];
}

- (void) testSoleArgUsingApplySyntax {
  ErrorBuffer *errors = [ErrorBuffer newErrorBuffer];
  NSString *templates = [@"t() ::= \"<{9}:u()>\"\n" stringByAppendingString:@"u(x,y) ::= \"<x>\"\n"];
  [self writeFile:tmpdir arg1:@"t.stg" arg2:templates];
  STGroup *group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"t.stg"]] autorelease];
  [group setListener:errors];
  ST *st = [group getInstanceOf:@"t"];
  NSString *expected = @"9";
  NSString *result = [st render];
  [self assertEquals:expected arg1:result];
  expected = [@"context [t] 1:5 passed 1 arg(s) to template u with 2 declared arg(s)" stringByAppendingString:newline];
  result = [errors description];
  [self assertEquals:expected arg1:result];
}

- (void) testUndefinedAttr {
  ErrorBuffer *errors = [ErrorBuffer newErrorBuffer];
  NSString *templates = [@"t() ::= \"<u()>\"\n" stringByAppendingString:@"u() ::= \"<x>\"\n"];
  [self writeFile:tmpdir arg1:@"t.stg" arg2:templates];
  STGroup *group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"t.stg"]] autorelease];
  [group setListener:errors];
  STGroup.debug = YES;
  ST *st = [group getInstanceOf:@"t"];
  [st render];
  NSString *expected = [@"context [t u] 1:1 attribute x isn't defined" stringByAppendingString:newline];
  NSString *result = [errors description];
  [self assertEquals:expected arg1:result];
}

- (void) testParallelAttributeIterationWithMissingArgs {
  ErrorBuffer *errors = [ErrorBuffer newErrorBuffer];
  STGroup *group = [[[STGroup alloc] init] autorelease];
  [group setListener:errors];
  ST *e = [[[ST alloc] init:group arg1:@"<names,phones,salaries:{n,p | <n>@<p>}; separator=\", \">"] autorelease];
  [e add:@"names" arg1:@"Ter"];
  [e add:@"names" arg1:@"Tom"];
  [e add:@"phones" arg1:@"1"];
  [e add:@"phones" arg1:@"2"];
  [e add:@"salaries" arg1:@"big"];
  [e render];
  NSString *errorExpecting = [[@"1:23: anonymous template has 2 arg(s) but mapped across 3 value(s)\n" stringByAppendingString:@"context [anonymous] 1:23 passed 3 arg(s) to template _sub1 with 2 declared arg(s)\n"] stringByAppendingString:@"context [anonymous] 1:1 iterating through 3 values in zip map but template has 2 declared arguments\n"];
  [self assertEquals:errorExpecting arg1:[errors description]];
  NSString *expecting = @"Ter@1, Tom@2";
  [self assertEquals:expecting arg1:[e render]];
}

- (void) testStringTypeMismatch {
  ErrorBuffer *errors = [ErrorBuffer newErrorBuffer];
  STGroup *group = [[[STGroup alloc] init] autorelease];
  [group setListener:errors];
  ST *e = [[[ST alloc] init:group arg1:@"<trim(s)>"] autorelease];
  [e add:@"s" arg1:34];
  [e render];
  NSString *errorExpecting = [@"context [anonymous] 1:1 function trim expects a string not java.lang.Integer" stringByAppendingString:newline];
  [self assertEquals:errorExpecting arg1:[errors description]];
}

- (void) testStringTypeMismatch2 {
  ErrorBuffer *errors = [ErrorBuffer newErrorBuffer];
  STGroup *group = [[[STGroup alloc] init] autorelease];
  [group setListener:errors];
  ST *e = [[[ST alloc] init:group arg1:@"<strlen(s)>"] autorelease];
  [e add:@"s" arg1:34];
  [e render];
  NSString *errorExpecting = [@"context [anonymous] 1:1 function strlen expects a string not java.lang.Integer" stringByAppendingString:newline];
  [self assertEquals:errorExpecting arg1:[errors description]];
}

@end

File: test/org/stringtemplate/v4/test/TestInterptimeErrors.java
/*
 [The "BSD license"]
 Copyright (c) 2009 Terence Parr
 All rights reserved.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:
 1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.
 2. Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.
 3. The name of the author may not be used to endorse or promote products
    derived from this software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
package org.stringtemplate.v4.test;

import org.junit.Test;
import org.stringtemplate.v4.*;
import org.stringtemplate.v4.misc.*;

import static org.junit.Assert.assertEquals;

public class TestInterptimeErrors extends BaseTest {
    public static class UserHiddenName {
        protected String name;
        public UserHiddenName(String name) { this.name = name; }
        protected String getName() { return name; }
    }

    public static class UserHiddenNameField {
        protected String name;
        public UserHiddenNameField(String name) { this.name = name; }
    }

    @Test public void testMissingEmbeddedTemplate() {
        ErrorBuffer errors = new ErrorBuffer();

        String templates =
            "t() ::= \"<foo()>\"" + Misc.newline;

        writeFile(tmpdir, "t.stg", templates);
        STGroup group = new STGroupFile(tmpdir+"/"+"t.stg");
        group.setListener(errors);
        ST st = group.getInstanceOf("t");
        st.render();
        String expected = "context [/t] 1:1 no such template: /foo"+newline;
        String result = errors.toString();
        assertEquals(expected, result);
    }

    @Test public void testMissingSuperTemplate() {
        ErrorBuffer errors = new ErrorBuffer();

        String templates =
            "t() ::= \"<super.t()>\"" + Misc.newline;

        writeFile(tmpdir, "t.stg", templates);
        STGroup group = new STGroupFile(tmpdir+"/"+"t.stg");
        group.setListener(errors);
        String templates2 =
            "u() ::= \"blech\"" + Misc.newline;

        writeFile(tmpdir, "t2.stg", templates2);
        STGroup group2 = new STGroupFile(tmpdir+"/"+"t2.stg");
        group.importTemplates(group2);
        ST st = group.getInstanceOf("t");
        st.render();
        String expected = "context [/t] 1:1 no such template: super.t"+newline;
        String result = errors.toString();
        assertEquals(expected, result);
    }

    @Test public void testNoPropertyNotError() {
        ErrorBuffer errors = new ErrorBuffer();

        String templates =
            "t(u) ::= \"<u.x>\"" + Misc.newline;

        writeFile(tmpdir, "t.stg", templates);
        STGroup group = new STGroupFile(tmpdir+"/"+"t.stg");
        group.setListener(errors);
        ST st = group.getInstanceOf("t");
        st.add("u", new User(32, "parrt"));
        st.render();
        String expected = "";
        String result = errors.toString();
        assertEquals(expected, result);
    }

    @Test public void testHiddenPropertyNotError() {
        ErrorBuffer errors = new ErrorBuffer();

        String templates =
            "t(u) ::= \"<u.name>\"" + Misc.newline;

        writeFile(tmpdir, "t.stg", templates);
        STGroup group = new STGroupFile(tmpdir+"/"+"t.stg");
        group.setListener(errors);
        ST st = group.getInstanceOf("t");
        st.add("u", new UserHiddenName("parrt"));
        st.render();
        String expected = "";
        String result = errors.toString();
        assertEquals(expected, result);
    }

    @Test public void testHiddenFieldNotError() {
        ErrorBuffer errors = new ErrorBuffer();

        String templates =
            "t(u) ::= \"<u.name>\"" + Misc.newline;

        writeFile(tmpdir, "t.stg", templates);
        STGroup group = new STGroupFile(tmpdir+"/"+"t.stg");
        group.setListener(errors);
        ST st = group.getInstanceOf("t");
        st.add("u", new UserHiddenNameField("parrt"));
        st.render();
        String expected = "";
        String result = errors.toString();
        assertEquals(expected, result);
    }

    @Test public void testSoleArg() {
        ErrorBuffer errors = new ErrorBuffer();

        String templates =
            "t() ::= \"<u({9})>\"\n"+
            "u(x,y) ::= \"<x>\"\n";

        writeFile(tmpdir, "t.stg", templates);
        STGroup group = new STGroupFile(tmpdir+"/"+"t.stg");
        group.setListener(errors);
        ST st = group.getInstanceOf("t");
        st.render();
        String expected = "context [/t] 1:1 passed 1 arg(s) to template /u with 2 declared arg(s)"+newline;
        String result = errors.toString();
        assertEquals(expected, result);
    }

    @Test public void testSoleArgUsingApplySyntax() {
        ErrorBuffer errors = new ErrorBuffer();

        String templates =
            "t() ::= \"<{9}:u()>\"\n"+
            "u(x,y) ::= \"<x>\"\n";

        writeFile(tmpdir, "t.stg", templates);
        STGroup group = new STGroupFile(tmpdir+"/"+"t.stg");
        group.setListener(errors);
        ST st = group.getInstanceOf("t");
        String expected = "9";
        String result = st.render();
        assertEquals(expected, result);

        expected = "context [/t] 1:5 passed 1 arg(s) to template /u with 2 declared arg(s)"+newline;
        result = errors.toString();
        assertEquals(expected, result);
    }

    @Test public void testUndefinedAttr() {
        ErrorBuffer errors = new ErrorBuffer();

        String templates =
            "t() ::= \"<u()>\"\n"+
            "u() ::= \"<x>\"\n";

        writeFile(tmpdir, "t.stg", templates);
        STGroup group = new STGroupFile(tmpdir+"/"+"t.stg");
        group.setListener(errors);
        ST st = group.getInstanceOf("t");
        st.render();
        String expected = "context [/t /u] 1:1 attribute x isn't defined"+newline;
        String result = errors.toString();
        assertEquals(expected, result);
    }

    @Test public void testParallelAttributeIterationWithMissingArgs() {
        ErrorBuffer errors = new ErrorBuffer();
        STGroup group = new STGroup();
        group.setListener(errors);
        ST e = new ST(group,
                "<names,phones,salaries:{n,p | <n>@<p>}; separator=\", \">"
            );
        e.add("names", "Ter");
        e.add("names", "Tom");
        e.add("phones", "1");
        e.add("phones", "2");
        e.add("salaries", "big");
        e.render();
        String errorExpecting =
            "1:23: anonymous template has 2 arg(s) but mapped across 3 value(s)" + newline +
            "context [anonymous] 1:23 passed 3 arg(s) to template /_sub1 with 2 declared arg(s)" + newline +
            "context [anonymous] 1:1 iterating through 3 values in zip map but template has 2 declared arguments" + newline;
        assertEquals(errorExpecting, errors.toString());
        String expecting = "Ter@1, Tom@2";
        assertEquals(expecting, e.render());
    }

    @Test public void testStringTypeMismatch()
    {
        ErrorBuffer errors = new ErrorBuffer();
        STGroup group = new STGroup();
        group.setListener(errors);
        ST e = new ST(group, "<trim(s)>");
        e.add("s", 34);
        e.render(); // generate the error
        String errorExpecting = "context [anonymous] 1:1 function trim expects a string not java.lang.Integer"+newline;
        assertEquals(errorExpecting, errors.toString());
    }

    @Test public void testStringTypeMismatch2()
    {
        ErrorBuffer errors = new ErrorBuffer();
        STGroup group = new STGroup();
        group.setListener(errors);
        ST e = new ST(group, "<strlen(s)>");
        e.add("s", 34);
        e.render(); // generate the error
        String errorExpecting = "context [anonymous] 1:1 function strlen expects a string not java.lang.Integer"+newline;
        assertEquals(errorExpecting, errors.toString());
    }
}
